File: 0062-unique-paths/0062-unique-paths.java
class Solution {
    public int uniquePaths(int m, int n) {
        int[] dp = new int[n];
        for(int i = 0;i<m;i++){
            int[] temp = new int[n];
            for(int j = 0;j<n;j++)
                if(i==0 && j == 0) temp[j] = 1;
                else{
                    int down = 0,right = 0;
                    if(i>0) down = dp[j];
                    if(j>0) right = temp[j-1];
                    temp[j] = down+right;
                }
            dp = temp;
        }
        return dp[n-1];
    }
} 

File: 0062-unique-paths/0062-unique-paths.cpp
class Solution {
public:
    bool check(int i, int j, int m, int n){
        if(i >= m || j >= n){
            return false;
        }
        return true;
    }
    int helper(int i, int j, int m, int n, vector<vector<int>> &dp){
        if(!check(i, j, m, n)){
            return 0;
        }
        if(i == m-1 && j == n-1){
            return 1;
        }
        if(dp[i][j] != -1){
            return dp[i][j];
        }
        return dp[i][j] = helper(i+1, j, m, n, dp) + helper(i, j+1, m, n, dp);
    }
    int uniquePaths(int m, int n) {
        int i = 0, j = 0;
        vector<vector<int>> dp(m, vector<int> (n, -1));
        return helper(i, j, m, n, dp);
    }
};